### PR TITLE
Add support for workflow jobs and steps

### PR DESCRIPTION
**Pull request by Tenzer**
This relates to GitHub Actions and #1373.

These new features expose information about the jobs run for a workflow, and each of the steps in those jobs.

I also made a small fix to the `add_attribute.py` script, as it added assertions that didn't work with datetime objects. I think it may have been leftover from the Python 2 support. Disabling the assertion instead made it work and also made the generated code match what I could see was used elsewhere in the project.

Please let me know if there's anything missing or any adjustments that need to be made.

### CODE REVIEW

File: github/WorkflowJob.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
+############################ Copyrights and license ############################
+#                                                                              #
+# Copyright 2020 Steve Kowalik <steven@wedontsleep.org>                        #
```
Comment (s-t-e-v-e-n-k) [inline github/WorkflowJob.py:3]: I did not write this file, please update this accordingly.

File: github/WorkflowStep.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
+############################ Copyrights and license ############################
+#                                                                              #
+# Copyright 2020 Steve Kowalik <steven@wedontsleep.org>                        #
```
Comment (s-t-e-v-e-n-k) [inline github/WorkflowStep.py:3]: As above.

File: scripts/add_attribute.py [hunk 3]
```diff
@@ -55,7 +55,7 @@
     ),
     "datetime": (
         "datetime.datetime",
-        "(str, unicode)",
+        None,
```
Comment (s-t-e-v-e-n-k) [inline scripts/add_attribute.py:58]: This is a useful change, but I may pull it out into a different PR

File: tests/WorkflowJob.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+############################ Copyrights and license ############################
+#                                                                              #
+# Copyright 2020 Steve Kowalik <steven@wedontsleep.org>                        #
```
Comment (s-t-e-v-e-n-k) [inline tests/WorkflowJob.py:3]: As above.

**Review by s-t-e-v-e-n-k — CHANGES_REQUESTED** [review]
Broadly this is looking good, just missing the type information (pyi files)

File: github/WorkflowJob.py (hunk 1 above)
Comment (Tenzer) [reply github/WorkflowJob.py]: Sorry about that. I copied the files when I made the new ones to make it easier to match the style and missed this. I have corrected it in 12af8b42f3a7e141d572c1cbe1cecf5ab38a067a.

File: github/WorkflowStep.py (hunk 2 above)
Comment (Tenzer) [reply github/WorkflowStep.py]: Fixed in 12af8b42f3a7e141d572c1cbe1cecf5ab38a067a.

File: tests/WorkflowJob.py (hunk 4 above)
Comment (Tenzer) [reply tests/WorkflowJob.py]: Fixed in 12af8b42f3a7e141d572c1cbe1cecf5ab38a067a.

File: scripts/add_attribute.py (hunk 3 above)
Comment (Tenzer) [reply scripts/add_attribute.py:58]: Would you like me to remove the commit with this change from this PR and make a new one, or do you want to take care of it?

**Comment by Tenzer** [comment]
Thanks for taking a look at this. I have added the missing type annotations now.

**Comment by seemethere** [comment]
Is there any update to this?

**Comment by Tenzer** [comment]
Is there any chance this could be incorporated?

**Comment by antoiner77** [comment]
Why is this PR closed ?

**Comment by crackedupcorson** [comment]
Is there any chance this feature could be added? I don't see WorkflowJob in this package currently.

**Comment by sfdye** [comment]
We need to resolve the merge conflicts

**Comment by Tenzer** [comment]
They have been sorted out now.

**Review by seemethere — APPROVED** [review]

**Comment by yongzhang** [comment]
can't wait it to be released

File: github/WorkflowJob.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        :type: string
+        """
+        self._completeIfNotSet(self._url)
+        return self._url.value
+
+    def logs_url(self):
+        """
+        :type: string
+        """
+        print(f"{self.url}/logs")
```
Comment (EnricoMi) [inline github/WorkflowJob.py:151]: I think this is not meant to be here.

File: github/WorkflowJob.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    @property
+    def url(self):
+        """
+        :type: string
+        """
+        self._completeIfNotSet(self._url)
+        return self._url.value
+
+    def logs_url(self):
```
Comment (EnricoMi) [inline github/WorkflowJob.py:147]: How come `WorflowRun.logs_url` is an ordinary string attribute?

https://github.com/PyGithub/PyGithub/blob/554b2b28db8e3c06915406cb10e39aceb38d16d7/github/WorkflowRun.py#L167-L173

What is different in `WorkflowJob`?

**Comment by EnricoMi** [comment]
@Tenzer I have added replay data for the logs_url call (and updated the other WorkflowJobs calls).

File: github/WorkflowJob.py (hunk 6 above)
Comment (Tenzer) [reply github/WorkflowJob.py:147]: I think `logs_url` is a field that has been added to the response since I originally wrote the code. I can see the captured response payload inside `tests/ReplayData/WorkflowRun.test_jobs.txt` doesn't have it, but the new one inside `tests/ReplayData/WorkflowJob.setUp.txt` does.

It can probably be changed to work the same way as `WorkflowRun` does.

**Review by davejames — APPROVED** [review]
I've done some functional testing on this and am currently using it. LGTM and hope it's merged soon

**Comment by FearlessHyena** [comment]
When will this be merged?

File: tests/WorkflowJob.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import datetime
+
+from . import Framework
+
+
+class WorkflowJob(Framework.TestCase):
+    def setUp(self):
+        super().setUp()
+        self.repo = self.g.get_repo("PyGithub/PyGithub")
+        self.job = self.repo.get_workflow_run(4205440316).jobs()[0]
```
Comment (EnricoMi) [inline tests/WorkflowJob.py:32]: Let's assert the number of jobs as well:
```suggestion
        jobs = self.repo.get_workflow_run(4205440316).jobs()
        self.assertEqual(len(jobs), 5)
        self.job = jobs[0]
```

File: github/WorkflowJob.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        """
+        :type: string
+        """
+        print(f"{self.url}/logs")
+        headers, _ = self._requester.requestBlobAndCheck("GET", f"{self.url}/logs")
+        return headers["location"]
```
Comment (EnricoMi) [inline github/WorkflowJob.py:153]: Let's get this in line with `WorkflowRun`:
```suggestion
    @property 
    def logs_url(self): 
        """ 
        :type: string 
        """ 
        self._completeIfNotSet(self._logs_url) 
        return self._logs_url.value 
```

File: github/WorkflowJob.py (hunk 6 above)
Comment (EnricoMi) [reply github/WorkflowJob.py:147]: Fix added as a suggestion.

**Review by EnricoMi — CHANGES_REQUESTED** [review]
Happy to approve once these suggestions are incorporated.

**Review by EnricoMi — APPROVED** [review]
LGTM!

**Review by sfdye — APPROVED** [review]
